### ISSUE TITLE
SDRAM timing

### ISSUE THREAD

**Issue by AVRsteffen** [Zigazou/myterminal#1]
First of all, thank you very much for this great project.  A very interesting implementation for a text generator with useful additions in my opinion.  I was particularly interested in the connection and use of SDRAM.  After a few modifications, I am now using the terminal for a 640x480 RGB TFT display with an 8-8-8 bit parallel RGB interface.  Unfortunately it caused problems.  The writing only started after the 3rd character and did not go through to the end.  It chewed up the last of the letters.  It cost me a lot of searching, but I was able to learn a lot about the interaction of the individual modules.  It was clearly a timing problem.  Found in the SDRAM module when reading the data.  The IDLE mode was jumped too early and so could not finish reading correctly.  The CAS_Latency time was missing when waiting to return to IDLE mode.

Modul: embedded_sdram.v
Task: stage_read
Line: 436
old: goto_when(STAGE_IDLE, read_burst_stop + 'd1 + tRP);
changed to: goto_when(STAGE_IDLE, read_burst_stop + CAS_LATENCY + tRP);

Now it works correct for me. But I think it is a generally problem. 
The second one I still search. The first data byte is always readed as hFF. I think its happen if will write data. But I am not sure.

**Comment by Zigazou** [comment]
Hi !

I would be glad to help you (please be kind because I'm a beginner).
Could you tell me more about the hardware you use ? Is it also a Tang Primer ?

It’s my very first FPGA project using SDRAM so I guess the timings aren’t perfect. I used this datasheet http://www.etron.com/manager/uploads/EM638325_Rev%203.2.pdf as it seems it is the kind of SDRAM the Tang Primer uses. On the Tang Primer, the memory controller seems quite stable, that’s why I’m asking these questions..

(Please note that the most advanced branch is newescapecodes.)

**Comment by Zigazou** [comment]
Hello !

The code looks like that because I’m a developper, not because I’m an FPGA expert ;-)

I will try to apply your changes and see what happens, thanks.

About the characters, if you don’t use the terminal_stream module and write values directly in memory, the four bytes used by each characters are organized as in this schema → https://github.com/Zigazou/myterminal/blob/master/doc/character_attributes.png

If you use the master version, I started with the idea of using UTF-8 and ANSI escape codes. But it turned out to be bandwidth consuming (UTF-8 for special characters are up to 3 bytes, ANSI escape sequences can be very long for simple things and you often lose 2 bytes juste to initiate an ANSI sequence) and resource consuming (converting UTF-8 to a custom 1024 character table for which lots of characters don't exist in UTF-8, interpreting ANSI escape sequences etc.).

Thus, my current move is to use dedicated escape codes (shorter and simpler to interpret) and ISO-8859-15 as base. The user sends only bytes and the page can be selected by escape codes.

The charset has been reorganised (compared to what you know), some characters have been added. If you open https://github.com/Zigazou/myterminal/blob/newescapecodes/font/extended_videotex.svg with Inkscape, you will be able to see the 5 character page available (characters 0x00 to 0x1F are always control codes thus not visible).

You can get the list of available escape codes in https://github.com/Zigazou/myterminal/blob/newescapecodes/src/terminal_stream/escape_codes.v.

There are some examples in the demo folder on how to use them. There even is a demo running in a Chrome browser allowing you to display things on MyTerminal using the Web Serial API.

I think using dedicated escape codes and ISO-8859-15 is more suited to small devices that could use this as a display.

Keep in mind this is work in progress and I can make strong changes if I think there is a better way ;-)

**Comment by Zigazou** [comment]
Yes, if you want to directly write characters in the video memory, you have to bypass terminal_stream. For a terminal, it’s more efficient to transmit only changes in the stream. Character attributes changes less often than character itself. Sending 4 bytes to write a new character using the same attributes than its predecessor is 4 times slower.

If you want to use the module more as a video card than as a terminal, you have to skip the terminal_stream module.

If you don't have a monitor or VGA converter, what do you use ?

COLUMNS_REAL is only there to simplify address calculation (it’s easier and needs less resources to multiply by 128 than by 80) and is not directly related to the SDRAM column size.

From what I read, the SDRAM column size is 256×32 bits not 256×8 bits.

**Comment by Zigazou** [comment]
Yes, if you want to directly write characters in the video memory, you have to bypass terminal_stream.

For a terminal, it’s more efficient to transmit only changes in the stream. Character attributes changes less often than characters itself. Sending 4 bytes to write a new character using the same attributes than its predecessor is 4 times slower.

If you want to use the module more as a video card than as a terminal, you have to skip the terminal_stream module and writes directly to SDRAM and send register values to the video_controller..

If you don't have a monitor or VGA converter, what do you use ?

COLUMNS_REAL is only there to simplify address calculation (it’s easier and needs less resources to multiply by 128 than by 80) and is not directly related to the SDRAM column size except for the fact that it uses less than 1024 bytes. From what I read, the SDRAM column size is 256×32 bits (256×4 bytes) not 256×8 bits, isn't it ?

**Comment by AVRsteffen** [comment]
> If you don't have a monitor or VGA converter, what do you use ?

I use a TFT display which has a 24bit RGB interface. It usually uses 8bit per color and Hsync, Vsync and another DataEnable (Visible) signal. I am currently only using the upper 3 bits of the RGB color byte. The display has only 640x480 pixels. But I still have some with 800x600 pixels. However, these have a 4-bit LVDS interface. I want to use it as a terminal for small AVR microcontrollers. That's why the chart replacement is so great.

Later I plan to adapt this to smaller QVGA (320x240 pixel) displays. However, I have to make the font smaller. So 8x10 or 10x14 char font. Otherwise the font will be too big.

I have already adapted and tried the New Escape Codes version for myself. Looks good - almost everything works. Great job Frédéric.

I don't understand how the patterns are generated yet. I send h03 as data and then?
Base h04 + what?
h10, h20, h30 .. for Pattern ID?
h01, h02, h03 .. for pattern offset?
gives basis h04 + h20 + h06 = h2A?

**Comment by Zigazou** [comment]
If you look for a smaller font, just choose 8×10 because that's my starting point : I used the font from the EF9345 chipset as a base that I upscaled.

There are 15 patterns, they are hardcoded and contained in the src/video_controller/generate_pattern.v. This is done to ensure that they can be repeated in horizontal or vertical directions without any glitch.

The pattern code sent after the h03 escape code is built as follows :
- the first 2 bits specify the function ID (4 possible values)
- the next 4 bits specify the pattern ID (16 possible values)
- the last 2 bits are set to 2'b01 (7th bit or bit[6] is set)

In Verilog pseudo-notation : { 2'b01, 4'b(pattern_id), 2'b(function_id) }
Or pattern_code = 0x40 + pattern_id * PATTERN_BASE + function_id
As FUNCTION_* constants already have the 0x40 part, it can be written FUNCTION_* + pattern_id * PATTERN_BASE (I admit PATTERN_BASE is not the best chosen name ;-) )
